### ISSUE TITLE
Test inbox pagination after loading all messages

### ISSUE THREAD

**Issue by sosnovsky** [FlowCrypt/flowcrypt-ios#1288]
Steps:

1. Login to e2e account
2. Observe Inbox screen - 36 emails are displayed
3. Pull down for refresh
4. Observe Inbox screen - it should load all 36 emails

Previously there was an issue when after loading all 36 emails and performing pull-to-refresh only 20 emails were loaded (https://github.com/FlowCrypt/flowcrypt-ios/issues/1284)

But emails count probably will change in the future, so test shouldn't depend on it - it'll be better to compare title of the last message on first load with last email title on the second load (after pull-to-refresh).

**Comment by fcvakintos** [comment]
@sosnovsky I get email count from inbox before scroll and after, and check that counts after scroll should be great than before

**Comment by sosnovsky** [comment]
Ah, great, then we should check if total email count after pull-to-refresh is the same as on the first load (currently it's 36 emails)

**Comment by fcvakintos** [comment]
@sosnovsky sure, how many email should be in Inbox screen for check pagination? 
I tried to check it in e2e account but 36 email in not enough

**Comment by sosnovsky** [comment]
Initial issue happened only after user have loaded all messages in Inbox. Is it possible to check when all emails have loaded?
Like this:
1) scroll down
2) check if number of emails changed
3) if changed - scroll down one more time to load next page of emails
4) if emails count didn't change - all messages are loaded, can do pull-to-refresh now

**Comment by fcvakintos** [comment]
For this case we have this ticket, https://github.com/FlowCrypt/flowcrypt-ios/issues/1156

but I dont know how many emails should be in inbox for pagination, with 36 emails pagination is not working, i think we need more emails

**Comment by fcvakintos** [comment]
I found 1 page contains 40 email

**Comment by sosnovsky** [comment]
I think currently we have 20 emails per page

**Comment by fcvakintos** [comment]
@sosnovsky hmmm. strange
I have 40 emails in page 
but after open some email/ click back to inbox screen-> the emails count = 20
check pls video
https://user-images.githubusercontent.com/91205853/148959052-f0c7ade8-62a6-423f-9cb6-ca834434bcc0.mov

**Comment by sosnovsky** [comment]
On `00:10` of your video you can see how page #2 of email is added on the bottom, as next page is automatically loaded when you scroll to the bottom of the list. So each page contains 20 emails which are automatically loaded on scroll.

But I just noticed another bug - emails shouldn't disappear when you go back from email detail screen. And also there is something wrong with sorting, as email with `Jan 10` date is below emails from November. I'll create separate issue for these issues

**Comment by sosnovsky** [comment]
So probably we should have such steps for inbox testing:
1) load inbox screen
2) check number of emails - it should be 20
3) scroll down - number of emails should increase
4) scroll down until number of emails not changing (it means all inbox emails are loaded)
5) perform pull-to-refresh - number of emails should be 20
6) scroll down - number of emails should increase

**Comment by tomholub** [comment]
wrong order filed here https://github.com/FlowCrypt/flowcrypt-ios/issues/1120 you can have a look at it while you are looking at the functionality. Order should match how it was received from Gmail API when pulling whole thread. Seems like my ordering fix didn't work in 100% of the cases.

**Comment by fcvakintos** [comment]
> So probably we should have such steps for inbox testing:
> 
> 1. load inbox screen
> 2. check number of emails - it should be 20
> 3. scroll down - number of emails should increase
> 4. scroll down until number of emails not changing (it means all inbox emails are loaded)
> 5. perform pull-to-refresh - number of emails should be 20
> 6. scroll down - number of emails should increase

ok, i will add this case

**Comment by sosnovsky** [comment]
> wrong order filed here #1120 you can have a look at it while you are looking at the functionality. Order should match how it's showing in Gmail webmail.
> 
> The response with list of inbox items we get from email should be ordered correctly (is it?)
> 
> * if it is, then it seems like my ordering fix earlier didn't work in 100% of the cases. We should fix that. Let's not do manual ordering by any criteria locally (eg date) other than ordering by the order it came from API originally.
> * if it's not, maybe we should add an `sort by date desc` of some sort to the API request

I think there is something on our side, as only 1 last message has wrong order - maybe inserting emails from loaded pages has some wrong calculation for messages index, need to check.

**Comment by tomholub** [comment]
@sosnovsky the status here now is that there is still 1 outstanding message that has wrong apparent date, caused as far as we know by Gmail API - is that correct?

**Comment by sosnovsky** [comment]
Currently last message has right date (it was fixed by https://github.com/FlowCrypt/flowcrypt-ios/commit/d4d799019d80d8400ef070e0b816658d0351e1d5), but several messages in the middle of the list has wrong dates (from 2017 and 2018).

**Comment by tomholub** [comment]
@fcvakintos please write the test anyway, and skip the messages that are known to have weird dates, in the test.

With the following comment:

```
/// Some messages have wrong dates returned from Gmail API, showing dates from 2017 or 2018 unexpectedly.
/// We are not evaluating these messages with wrong dates in this test until it's resolved.
```

**Comment by fcvakintos** [comment]
@sosnovsky what is default count of emails on inbox screen?
20 or 40?
I see 40 before pagination

**Comment by sosnovsky** [comment]
Each page has 20 emails, you see 40 emails because they're automatically loaded during scroll

**Comment by fcvakintos** [comment]
@sosnovsky I see 20 after login without scroll

**Comment by fcvakintos** [comment]
@sosnovsky maybe is it a bug?

**Comment by sosnovsky** [comment]
Yes, without scroll you should see 20 emails, then when you scroll to the bottom of the list app should load next page with 20 emails. Does it load page 2 for you?

**Comment by fcvakintos** [comment]
@sosnovsky yes, from 40 to 45 email

**Comment by fcvakintos** [comment]
if i open some email and back to inbox - count = 20 without scroll

**Comment by sosnovsky** [comment]
That's right, so there is no bug there. Then test should looks like this:
1. open Inbox screen
2. check number of messages (should be 20)
3. scroll to the bottom to load next page
4. check number of messages (should be more than 20)
5. scroll to the top of the list and perform pull-to-refresh
6. check number of messages (should be 20)
7. scroll to the bottom to load next page
8. check number of messages (should be more than 20)

Steps 2-4 and 6-8 are the same, so you can move them to helper method `checkPagination`

**Comment by sosnovsky** [comment]
> if i open some email and back to inbox - count = 20 without scroll

It's a bug, we have an issue for it - https://github.com/FlowCrypt/flowcrypt-ios/issues/1289

**Comment by fcvakintos** [comment]
so default is 40?

**Comment by sosnovsky** [comment]
1 page = 20 emails

**Comment by fcvakintos** [comment]
my steps:

-  login -> Inbox screen is open
- check count - > 40 emails 
- scroll down to the first email on the inbox screen
- check count - > 45 emails 
- scroll up to last email on inbox 
- refresh screen - > 40 emails are displayed
- open some email
- click back button - >inbox screen is open
- check count  - > 20 emails are displayed

So i think bug is -  the default emails count after first login  = 40

**Comment by fcvakintos** [comment]
and  
- open setting 
- open back inbox screen
count of emails = 40

**Comment by fcvakintos** [comment]
@sosnovsky can you check in e2e.enterprise.test@flowcrypt.com?

**Comment by sosnovsky** [comment]
checking it

**Comment by fcvakintos** [comment]
thanks

**Comment by sosnovsky** [comment]
I reproduced this issue on my side - app loads page 2 even without scroll, I'll add this bug to issue https://github.com/FlowCrypt/flowcrypt-ios/issues/1366

So for now please check if initial count is 40 emails (not 20) and after scrolling to the bottom of the list it should be more than 40. And also add a comment there saying `Should be changed to 20 when https://github.com/FlowCrypt/flowcrypt-ios/issues/1366 is fixed`

**Comment by fcvakintos** [comment]
@sosnovsky thanks
